### DISCUSSION TITLE
Typescript에서 axios post 요청에 대한 response data type을 지정해주는 방법

### DISCUSSION THREAD

**Discussion by HyeongtaekOh** [wafflestudio/seminar-2021#634] [Q&A] [answered]
제목 그대로입니다. tsx 마이그레이션을 하려는데 post 요청의 경우 response data에 generic을 어떻게 작성해줘야 하는지 모르겠습니다. 
지난 세미나에서 직접 보여주신 과정을 참고하고 있는데 해당 부분에서는 `axios.get<Tdata>~~` 이런 식으로 적어주신걸로 기억하는데 get request data가 따로 없어서 이렇게 적어도 해당 request의 response data에 대한 generic type으로 들어간 것인가요?
유사하게 `axios.post<DataType>~~`와 같은 식으로 코드를 작성했더니 해당 generic type이 request data의 type으로 인식이 되는 것 같은데 어떤 식으로 해야할지 궁금합니다.

**Comment by woohm402** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/39977696/140468836-61394eab-507c-4ca9-837e-77d52612d3e4.png)

전 이렇게 잘 뜨는데 혹시 어떻게 뜨시나요?

**Reply by HyeongtaekOh** [reply]
![스크린샷, 2021-11-05 15-47-58](https://user-images.githubusercontent.com/86342741/140469791-68bb0677-5efd-45a5-a633-41804652d94f.png)
![스크린샷, 2021-11-05 15-48-55](https://user-images.githubusercontent.com/86342741/140469864-92fd68ce-3606-4ea7-97dd-384155c6c3bf.png)

**Reply by HyeongtaekOh** [reply]
학생 등록할 때 request data type으로 들어가있습니다

**Reply by HyeongtaekOh** [reply]
`ResponseData` type은 `{ success: boolean; id: number; }`입니다

**Reply by woohm402** [reply]
아 제가 구버전을 쓰고 있었네요 (`0.18.0`) 확인해보니 `0.22.0` 사용중이신 것 같습니다.

공식문서 확인해 보시면 ([소스](https://github.com/axios/axios/blob/c5fe05bdff82e0fa4ace9a832e75052e1ee297f0/index.d.ts#L182))

버전 `0.22.0` 에서 axios의 post함수는 타입이 아래와 같이 잡혀 있습니다.

```ts
export class Axios {
  ...
  post<T = any, R = AxiosResponse<T>, D = any>(url: string, data?: D, config?: AxiosRequestConfig<D>): Promise<R>;
  ...
}
```

그래서 해당 버전에서 `post` 의 리스폰스 타입을 잡아주려면 제네릭 인자를 두개 줘야 한다고 합니다. 두 번째 인자를 안 주면, 자동으로 `AxiosResponse<리퀘스트 타입>` 으로 잡히겠군요

```ts
axios.post<RequestData, AxiosResponse<ResponseData>>('...
```

**Reply by HyeongtaekOh** [reply]
오 해결됐습니다!
